Upgrade to Lapack lite 3.2.2

eric-wieser opened the pull request:
Implements #8376. We pick the earliest acceptable version for the bug this was aiming to fix ([`dgelsd` does not correctly compute `iwork`](http://icl.cs.utk.edu/lapack-forum/archives/lapack/msg00899.html)), because each new version throws more fortran features not supported by `f2c` at us.

This is fully functional<s>, but a little hacky right now.</s>

Not sure  how best to structure the commits here. Unlike the last PR, these don't stand up by themselves - but I don't really want to stick a huge generated diff in the same place as hand-written code.

eric-wieser committed:
ENH: Add patch files to allow LAPACK 3.2.2 to be f2c'd

This doesn't yet actually generate the files, since they would cause the diff to balloon

eric-wieser commented on numpy/linalg/lapack_lite/dlapack_lite.f.patch:
```diff
@@ -0,0 +1,239 @@
+-                  END DO
++C                 DO LASTV = N, I+1, -1
++C                    IF( V( LASTV, I ).NE.ZERO ) EXIT
++C                 END DO
```
This is a pretty bad patch, that instead of actually rewriting this code in F77 style, just says "oh, we probably didn't need this anyway"

homu commented:
:umbrella: The latest upstream changes (presumably #8650) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
Given the huge number of `-Wparentheses` warnings in the test, it would be worth disabling that warning for gcc, see http://stackoverflow.com/questions/3378560/how-to-disable-gcc-warnings-for-a-few-lines-of-code. That should be OK for generated code where human readability for review is less a factor. I'd suggest the same for other compilers, but don't know if they actually generate relevant warning by default. Or might be worth checking if there is a less stringent warning flag for the tests.

eric-wieser commented:
Yep, previously we used a patched `f2c` that added the parentheses, but that patch was lost to time, and forcing numpy developers to recompile f2c is a bit mean.

This problem was actually present in #8381 - should have made more noise there!

I'll make a quick Pr for that at some poitn

charris commented:
I didn't see it when I compiled on my machine with `-Wall -Wstrict-prototypes`. I don't think we should apply fixes throughout, but disabling the check for the generated code makes sense.

eric-wieser commented:
Hmm, I can't get travis to show me the warnings for that build either. Perhaps it only shows warnings if any errors occur?

charris commented:
I'm wondering if pip picks up CFLAGS. However, I see that you introduced a new warning in 564b7cbc
```
numpy/core/src/multiarray/shape.c:708:39: warning: passing argument 1 of ‘check_and_adjust_axis’ from incompatible pointer type [-Wincompatible-pointer-types]
```

eric-wieser commented:
@charris: Hmm, good spot. I was tempted to use `npy_intp` for the axis parameter, but I noticed that a bunch of places already used `int`, and kind of assumed they all did

eric-wieser committed:
ENH: Add more functions copied to f2c_lite.

These are copied from the libf2c source code, which is presumably where all
the others came from

eric-wieser committed:
ENH: Rebuild all of lapack_lite from 3.2.2

eric-wieser commented on numpy/linalg/lapack_lite/lapack.f.patch:
```diff
@@ -0,0 +1,56 @@
+-            END DO
++C           DO I = M, 1, -1
++C              IF( A(I, J).NE.ZERO ) EXIT
++C           END DO
```
This is a pretty bad patch, that instead of actually rewriting this code in F77 style, just says "oh, we probably didn't need this anyway". The entire set of patches is rewriting this kind of loop.

eric-wieser replied:
[There's a patch here](https://github.com/igraph/igraph/blob/0f41313882a5a491a1bfb0d26a52feee569734ce/tools/lapack/lapack.patch#L8-L19) that probably better fixes this, which I should probably copy here.

mhvk replied:
@eric-wieser - I was very surprised that this wouldn't just convert. Trying bit with `f2c` (as packaged by debian testing), the problem does not seem to be the `do, enddo` construction, but rather the `exit` statement.  It should not be needed to do more than replace that with a `goto <end-label>`.

p.s. I wish we could just compile the fortran; `gfortran` is a very nice compiler...

eric-wieser replied:
I can certainly give that a go - thanks @mhvk

mhvk replied:
Well, if the other patch just works, then maybe that is easier. Anyway, let me know if there are issues; I'm old enough for most of my scientific code to have been written in fortran.

eric-wieser commented:
Build cancelled on travis, since this will be rebased anyway

eric-wieser commented:
There's a [now-hidden conversation here](https://github.com/numpy/numpy/pull/8649#discussion_r103826586) about the contents of the `.patch` files, that I should draw attention to.

charris commented:
@eric-wieser What is the status of this?

eric-wieser commented:
@charris: Fortran patches _probably_ need work, and I haven't found the time to improve them. It's possible the bits of code in need of patching are just optimizations though, and that everything is working correctly right now.

eric-wieser commented:
@charris: Patches are fixed as recommended by @mhvk

charris commented on numpy/linalg/lapack_lite/f2c_lapack.f.patch:
```diff
@@ -0,0 +1,48 @@
+-               IF( A(I, J).NE.ZERO ) EXIT
+-            END DO
++               IF( A(I, J).NE.ZERO ) GOTO 10
++   10       END DO
```
I don't think this is correct for the EXIT, what you have here is equivalent to the C continue, whereas you want a break, i.e., GOTO the first statement after the END DO. See https://gcc.gnu.org/onlinedocs/gcc-3.4.4/g77/CYCLE-and-EXIT.html .

eric-wieser replied:
Looking at the generated code, I can confirm you're correct. Drat.

charris commented:
I don't think the EXIT fixes are correct.

eric-wieser commented:
Alarming that tests didn't catch that.

eric-wieser commented:
Ok, I've added a pair of fixup commits that resolve that. If they look ok, then I'll rebase them in

charris commented on numpy/linalg/lapack_lite/README.rst:
```diff
@@ -27,10 +27,10 @@ similar to that done to generate the CLAPACK_ distribution.
+The output C files in git use the LAPACK source from the LAPACK_ page, using
+version 3.2.2. Unfortunately, the newer version use newer and newer FORTRAN
+features that are increasingly not supported bt ``f2c``. As these are found,
+the patch files need to be changed
```
'will need'

charris commented on numpy/linalg/lapack_lite/README.rst:
```diff
@@ -27,10 +27,10 @@ similar to that done to generate the CLAPACK_ distribution.
-released version of LAPACK available at the LAPACK_ page.
+The output C files in git use the LAPACK source from the LAPACK_ page, using
+version 3.2.2. Unfortunately, the newer version use newer and newer FORTRAN
+features that are increasingly not supported bt ``f2c``. As these are found,
```
'bt' <- 'by'. Maybe just one "newer"

charris replied:
'version' <- 'versions'

charris commented on numpy/linalg/lapack_lite/f2c_config.c:
```diff
@@ -107,7 +106,6 @@ doublereal dlamch_(char *cmach)
 
 
     if (first) {
-	first = FALSE_;
```
Any particular reason to move this?

eric-wieser replied:
Because the oracle that is `f2c` decided to. This is generated code beyond my control

charris commented on numpy/linalg/lapack_lite/f2c_config.c:
```diff
@@ -239,7 +237,6 @@ doublereal dlamch_(char *cmach)
 
 
     if (first) {
-	first = FALSE_;
```
See above. I think it is easier to follow setting `first` here, i.e., execute only the first time.

charris replied:
And so on for the rest.

charris commented on numpy/linalg/lapack_lite/make_lite.py:
```diff
@@ -23,7 +23,7 @@
 # Others of interest: -a to not make variables static by default
 #                     -C to check array subscripts
-F2C_ARGS = ['-A']
+F2C_ARGS = ['-A', '-Nx800']
```
What is the `-N` option?

eric-wieser replied:
Something to do with function table sizes. It might no longer be needed now that dlapack is a little smaller

mhvk commented on numpy/linalg/lapack_lite/f2c_c_lapack.f.patch:
```diff
@@ -0,0 +1,32 @@
+-                        IF( V( LASTV, I ).NE.ZERO ) EXIT
++                        IF( V( LASTV, I ).NE.ZERO ) GO TO 35
+                      END DO
++   35                CONTINUE
```
I'd put the label on the next statement, but really, that is total nitpicking. This certainly works...

eric-wieser replied:
This is the whole point in `CONTINUE` though, right?

Did this because it makes the patch more self-contained

mhvk replied:
ok, that makes sense.

charris replied:
Ah, another old Fortran guy. I remember when the compiler was written in assembly, came on paper tape, and ran in several kilobytes of memory. Not only that, but a single human being could read the code and understand it.

mhvk replied:
Haven't used paper type, but did use punch cards...  But I fear my real early programming experience was in BASIC on a PET 2001...

charris replied:
Ah, different generation. The good thing about paper tape was that you could edit it with a razor blade and scotch tape 8-)

Anyway, I confess that the first computer I programmed for was one of the bigger ones in the country at that time, an IBM 360 with two megabytes of memory.

charris commented:
Patches look good, couple of other comments. Why the big increase in size? More functions or more complex code?

eric-wieser commented:
> Why the big increase in size? 

I think a lot of the documentation got better. Remember that we're looking at 10 years of development to LAPACK here

charris commented:
LGTM. Could you add something to the 1.13 release notes?

charris commented:
I think you could just add your name and a date. If anyone wants to find out more they can use git blame to track it back to the relevant commit.

charris commented:
Or you could just leave it be. Pearu's name is still in some of the f2py files as an historical artifact even though others have worked on the source since.

eric-wieser committed:
DOC: Add changelog entry for new lapack_lite

Also, update authors of lapack_lite

[ci skip]

charris commented:
In it goes. Thanks Eric.
